lib/loglimit/loglimit.go

```diff
@@ -0,0 +1,205 @@
+	LogSubstrings []string
+	// ChannelSize is the size of the channel used to send messages
+	// to the log limiter.
+	ChannelSize int
+	// Clock is a clock to override in tests, set to real time clock
+	// by default.
+	Clock clockwork.Clock
+}
+
+// checkAndSetDefaults verifies configuration and sets defaults
+func (c *Config) checkAndSetDefaults() error {
+	if c.LogSubstrings == nil {
+		return trace.BadParameter("missing parameter LogSubstrings")
+	}
+	if c.ChannelSize < 0 {
+		return trace.BadParameter("ChannelSize must be at least 0")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// LogLimiter deduplicates logs over a certain time window.
+type LogLimiter struct {
+	Config
+	// entryCh is used to send log entries to the log limiter.
+	entryCh chan entryInfo
+	// windows is a mapping from log substring to an active
+	// time window.
+	windows map[string]*windowInfo
+}
+
+// entryInfo contains information about a certain log entry.
+type entryInfo struct {
+	// entry is the logger entry.
+	entry *log.Entry
+	// level is the log level at which this log entry should be logged.
+	level log.Level
+	// message is the message message.
+	message string
+	// time is the at which the log reported.
+	time time.Time
+}
+
+// windowInfo contains the information necessary to deduplicate a log
+// entry reported within a certain time window.
+type windowInfo struct {
+	// entryInfo contains information about the first log entry
+	// reported for this window such that entryInfo.time represents
+	// the beginning of this time window.
+	entryInfo
+	// occurrences are the occurrences of logs entries (that share
+	// the same log substring) within this window.
+	occurrences int
+}
+
+// New creates a new log limiter.
+func New(cfg Config) (*LogLimiter, error) {
+	if err := cfg.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &LogLimiter{
+		Config:  cfg,
+		entryCh: make(chan entryInfo, cfg.ChannelSize),
+		windows: make(map[string]*windowInfo, len(cfg.LogSubstrings)),
+	}, nil
+}
+
+// Log sends a log to the log limiter.
+func (l *LogLimiter) Log(entry *log.Entry, level log.Level, args ...any) {
+	l.entryCh <- entryInfo{
+		entry:   entry,
+		level:   level,
+		message: sprintln(args...),
+		time:    l.Clock.Now(),
+	}
+}
+
+// Run runs the log limiter.
+func (l *LogLimiter) Run(ctx context.Context) {
+	t := l.Clock.NewTicker(timeWindowCleanupInterval)
+	defer t.Stop()
+
+	for {
+		select {
+		case e := <-l.entryCh:
+			l.deduplicate(e)
+		case <-t.Chan():
+			l.cleanup()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// deduplicate logs if they should not be deduplicated.
+// Otherwise, it records log occurrences within a certain time window.
+func (l *LogLimiter) deduplicate(e entryInfo) {
+	// Log right away if the log entry should not be deduplicated.
+	deduplicate, logSubstring := l.shouldDeduplicate(&e)
+	if !deduplicate {
+		e.entry.Logln(e.level, e.message)
```

Comment:
Suggestion: Use `e.entry.Log`, so the caller can decide on whether to add newlines of not. I don't think we usually use "ln" variants for logging, instead we let the framework deal with it.

Same for others.

Reply:
Done in a9299e3e388e37a01b24e96b4f5fcc01ce5f3752.